### ISSUE TITLE
Compiling mcs

### ISSUE THREAD

**Issue by z3t0** [aeroson/mcs-ICodeCompiler#5]
Hi,

I'd like to compile mcs myself and was wondering if you could add some more details as to how you compiled it. Such as what was removed from the Mono SDK and then what externals were added. 

Thank you :smile:

**Comment by z3t0** [comment]
For example why is ikvm needed when, afaik, this has to do with providing a Java vm. Which I would assume is not needed for unity ?

**Comment by JakubNei** [comment]
I looked at the missing files included in mcs.csproj and downloaded them. Those are the missing external repositories (or submodules).

MCS can be compiled either for STATIC or DYNAMIC compilation.
STATIC compilation uses parts of IKVM (IKVM.Reflection) and avoids System.Reflection.Emit.
DYNAMIC compilation uses System.Reflection.Emit.

DYNAMIC creates the assembly in memory and I think it thus can not be easily unloaded. But once compilation is finished you already have it loaded.

STATIC you probably need to save & load the assembly.

**Comment by z3t0** [comment]
Thank you for the very detailed response. 

One more question. If using static, would it then create at .dll which would then need to be loaded? 

Is there some where I can find more documentation about compiling mcs by itself?

**Comment by JakubNei** [comment]
Probably yes, but you can do the same with DYNAMIC compilation as well.
It's just on some platforms DYNAMIC compilation (System.Reflection.Emit) might not be available.

All the information is in the code, build scripts, etc.. : https://github.com/mono/mono
I just took my time and understood what was needed, plus alot of trial and errors.

**Comment by z3t0** [comment]
Awesome thank you!

**Comment by z3t0** [comment]
At your convenience could you also please explaining the following.

> Find a way to run jay (the parser generator), mostly from looking at the code of it and or the Makefiles
> Jay parser generator was compiled and then ran using the mcs/jay/#_GENERATE_PARSER_FROM_cs-parser.jay.bat
> Once jay is used, the cs-parser.jay is transformed into parser file called cs-parser.cs
- I am using Windows so would I have to install jay somehow?
- How would I go about compiling jay? Is it the jay folder within mono/mcs
- Do you know how large the MonoSDK would be without stripping it down to just mcs? (Just out of curiosity)

**Comment by z3t0** [comment]
Okay so I have figured out that jay is provided in the mono folder and can be compiled, so that was kind of a stupid question... 

And it seems the Mono runtime provided with Unity is ~100 MB so downsizing to ~5 MB is definitely a win.

Also, in terms of flexibility and doing things 'they way they are supposed to' would you say that your method is better than simply running gmcs from the Mono Runtime? I'm curious as to which method is better, if the file sizes are not regarded.

**Comment by z3t0** [comment]
Okay for real this time last question :wink:

> Compile mcs.dll with .NET subset for Unity provided by Microsoft Visual Tools for Unity

How do I choose to use the .NET Unity subset within visual studio?

**Comment by z3t0** [comment]
> Change all internal classes to public, so they can be used in modified driver (driver = main class of mcs).

The file is about ~20 000 lines.... is there any automated way to achieve this?

**Comment by JakubNei** [comment]
> Also, in terms of flexibility and doing things 'they way they are supposed to' would you say that your method is better than simply running gmcs from the Mono Runtime? I'm curious as to which method is better, if the file sizes are not regarded.

In some scenarios my way could be faster. Because the gmcs.exe requires files on disc and generates result files back to disc. You can also do anything the code allows you to do, in gmcs.exe you are limited by the parameters. There is probably some extra overhead in the start of new process (gmcs).

> How do I choose to use the .NET Unity subset within visual studio?

You can download it here: [Visual Studio 2015 Tools for Unity](https://visualstudiogallery.msdn.microsoft.com/8d26236e-4a64-4d64-8486-7df95156aba9)
Or during Visual Studio 2015 installation you probably have an option to install it.

![image](https://cloud.githubusercontent.com/assets/3437520/16614629/8f0f846c-4372-11e6-8a4c-a0a69e938ddf.png)

> The file is about ~20 000 lines.... is there any automated way to achieve this?

Find & replace all private/internal class to public class.
![image](https://cloud.githubusercontent.com/assets/3437520/16614576/576412a8-4372-11e6-9f3c-e5b44c6ddcca.png)

**Comment by z3t0** [comment]
Thanks again!
